llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -265,6 +281,76 @@ void collectCompositeElementsInfoRecursive(
   }
 }
 
+/// Recursively iterates over a composite type in order to collect information
+/// about its scalar elements.
+void collectCompositeElementsDefaultValuesRecursive(
+    const Module &M, Constant *C, unsigned &Index, unsigned &Offset,
+    std::vector<char> &DefaultValues) {
+  Type *Ty = C->getType();
+  if (auto *ArrTy = dyn_cast<ArrayType>(Ty)) {
+    for (size_t I = 0; I < ArrTy->getNumElements(); ++I) {
+      Constant *El = cast<Constant>(C->getOperand(I));
+      // TODO: this is a spot for potential optimization: for arrays we could
+      // just make a single recursive call here and use it to populate
+      // DefaultValues in a loop.
```

Comment:
I'm not sure that this `TODO` is actual here: it is true that optimization is possible when you calculate offsets to each element of array: they are basically known if you know the size of an element. But for default values this can't be done without iterating over each element, as each element can have different value.

The same comment applies to handling of vectors